lib/chef/provider/systemd_unit.rb

```diff
@@ -0,0 +1,172 @@
+#
+# Author:: Nathan Williams (<nath.e.will@gmail.com>)
+# Copyright:: Copyright 2016, Nathan Williams
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/provider"
+require "chef/mixin/which"
+require "chef/mixin/shell_out"
+
+class Chef
+  class Provider
+    class SystemdUnit < Chef::Provider
+      include Chef::Mixin::Which
+
+      provides :systemd_unit, os: "linux"
+
+      def load_current_resource
+        @current_resource = Chef::Resource::SystemdUnit.new(new_resource.name)
+
+        current_resource.user(new_resource.user)
+        current_resource.enabled(enabled?)
+        current_resource.active(active?)
+        current_resource.masked(masked?)
+        current_resource.static(static?)
+
+        current_resource
+      end
+
+      def action_enable
+        if current_resource.static
+          Chef::Log.warn("#{new_resource.name} is a static unit, enabling is a NOP.")
+        end
+
+        if current_resource.enabled || current_resource.static
+          new_resource.updated_by_last_action(false)
+        else
+          systemctl_execute!(:enable, new_resource.name)
+          new_resource.updated_by_last_action(true)
+        end
```

Comment:
The manual twiddling of updated_by_last_action isn't necessary and eventually points to a why-run bug that you've got with this code.

systemctl_execute calls shell_out_with_systems_locale! directly without wrapping in a converge_by so there's no protection in why-run mode against running that code.

you also don't need to set updated_by_last_action since converge_by does that automatically, and false is the default value.  

so code should always look like:

```ruby
   if some_idempotent_condition?
    converge_by("do a thing") do    # will output "would do a thing" in why-run mode and then no-ops the block
       shell_out!("/usr/bin/do a thing")
    end
  end
```

generally code should never manually twiddle updated_by_last_action any more.

of course that statement relies on the fact that i just added use_inline_resources into Chef::Provider last week -- you'll still see a lot of core chef code which declares a sub-resource and manually copies off the updated status onto the new resource, but all of that can be converted now (and should be converted eventually) to use_inline_resources.